# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/01/04 19:48:45+01:00 trond.myklebust@fys.uio.no 
#    Subject: [PATCH] NFS: incorrect "df" results
#   
#    Description:
#    Fix an NFS client bug introduced in 2.6.9-rc1.  The "df" command was
#    reporting the size of NFS file systems incorrectly.
#   
#    Signed-off-by: Chuck Lever <cel@netapp.com>
#    Signed-off-by: Trond Myklebust <trond.myklebust@fys.uio.no>
# 
# fs/nfs/inode.c
#   2005/01/04 01:00:00+01:00 trond.myklebust@fys.uio.no +15 -1
#    Subject: [PATCH] NFS: incorrect "df" results
# 
diff -Nru a/fs/nfs/inode.c b/fs/nfs/inode.c
--- a/fs/nfs/inode.c	2005-01-04 19:59:27 -08:00
+++ b/fs/nfs/inode.c	2005-01-04 19:59:27 -08:00
@@ -486,13 +486,27 @@
 	if (error < 0)
 		goto out_err;
 
-	buf->f_frsize = server->wtmult;
+	/*
+	 * Current versions of glibc do not correctly handle the
+	 * case where f_frsize != f_bsize.  Eventually we want to
+	 * report the value of wtmult in this field.
+	 */
+	buf->f_frsize = sb->s_blocksize;
+
+	/*
+	 * On most *nix systems, f_blocks, f_bfree, and f_bavail
+	 * are reported in units of f_frsize.  Linux hasn't had
+	 * an f_frsize field in its statfs struct until recently,
+	 * thus historically Linux's sys_statfs reports these
+	 * fields in units of f_bsize.
+	 */
 	buf->f_bsize = sb->s_blocksize;
 	blockbits = sb->s_blocksize_bits;
 	blockres = (1 << blockbits) - 1;
 	buf->f_blocks = (res.tbytes + blockres) >> blockbits;
 	buf->f_bfree = (res.fbytes + blockres) >> blockbits;
 	buf->f_bavail = (res.abytes + blockres) >> blockbits;
+
 	buf->f_files = res.tfiles;
 	buf->f_ffree = res.afiles;
 
